File: mobile/src/utils/permissions.ts
/**
 * Role-Based Permission Helper for React Native
 * Centralizes permission checking logic for the mobile app
 */

export type UserRole = 
  | 'ADMIN'
  | 'MANAGER'
  | 'FRONT_DESK'
  | 'HOUSEKEEPING'
  | 'MAINTENANCE'
  | 'ACCOUNTANT'
  | 'POS_STAFF'
  | 'GUEST';

export interface User {
  id: number;
  email: string;
  first_name: string;
  last_name: string;
  role: UserRole;
  is_superuser?: boolean;
  assigned_property?: number | null;
}

/**
 * Check if user is superuser
 */
export const isSuperuser = (user: User | null): boolean => {
  return user?.is_superuser === true;
};

/**
 * Check if user has Admin or Manager role
 */
export const isAdminOrManager = (user: User | null): boolean => {
  if (!user) return false;
  return isSuperuser(user) || ['ADMIN', 'MANAGER'].includes(user.role);
};

/**
 * Check if user has Front Desk or above role
 */
export const isFrontDeskOrAbove = (user: User | null): boolean => {
  if (!user) return false;
  return isSuperuser(user) || ['ADMIN', 'MANAGER', 'FRONT_DESK'].includes(user.role);
};

/**
 * Check if user has Housekeeping staff role
 */
export const isHousekeepingStaff = (user: User | null): boolean => {
  if (!user) return false;
  return isSuperuser(user) || ['ADMIN', 'MANAGER', 'HOUSEKEEPING'].includes(user.role);
};

/**
 * Check if user has Maintenance staff role
 */
export const isMaintenanceStaff = (user: User | null): boolean => {
  if (!user) return false;
  return isSuperuser(user) || ['ADMIN', 'MANAGER', 'MAINTENANCE'].includes(user.role);
};

/**
 * Check if user has Accountant or above role
 */
export const isAccountantOrAbove = (user: User | null): boolean => {
  if (!user) return false;
  return isSuperuser(user) || ['ADMIN', 'MANAGER', 'ACCOUNTANT', 'FRONT_DESK'].includes(user.role);
};

/**
 * Check if user has POS staff role
 */
export const isPOSStaff = (user: User | null): boolean => {
  if (!user) return false;
  return isSuperuser(user) || ['ADMIN', 'MANAGER', 'POS_STAFF'].includes(user.role);
};

/**
 * Check if user can manage properties
 */
export const canManageProperties = (user: User | null): boolean => {
  return isSuperuser(user);
};

/**
 * Check if user can manage users
 */
export const canManageUsers = (user: User | null): boolean => {
  return isSuperuser(user);
};

/**
 * Navigation screen permission mapping
 */
export interface NavScreen {
  name: string;
  label: string;
  icon?: string;
  permission: (user: User | null) => boolean;
}

/**
 * Define all navigation screens with their required permissions
 */
export const navigationScreens: NavScreen[] = [
  {
    name: 'Dashboard',
    label: 'Dashboard',
    permission: (user) => isFrontDeskOrAbove(user) || isAdminOrManager(user),
  },
  {
    name: 'Properties',
    label: 'Properties',
    permission: (user) => isSuperuser(user) || isAdminOrManager(user),
  },
  {
    name: 'Users',
    label: 'Users',
    permission: (user) => canManageUsers(user),
  },
  {
    name: 'Reservations',
    label: 'Reservations',
    permission: (user) => isFrontDeskOrAbove(user),
  },
  {
    name: 'Guests',
    label: 'Guests',
    permission: (user) => isFrontDeskOrAbove(user),
  },
  {
    name: 'Rooms',
    label: 'Rooms',
    permission: () => true, // All authenticated users
  },
  {
    name: 'Housekeeping',
    label: 'Housekeeping',
    permission: (user) => isHousekeepingStaff(user),
  },
  {
    name: 'Maintenance',
    label: 'Maintenance',
    permission: (user) => isMaintenanceStaff(user),
  },
  {
    name: 'Billing',
    label: 'Billing',
    permission: (user) => isAccountantOrAbove(user),
  },
  {
    name: 'POS',
    label: 'POS',
    permission: (user) => isPOSStaff(user),
  },
  {
    name: 'Reports',
    label: 'Reports',
    permission: (user) => isAdminOrManager(user),
  },
];

/**
 * Filter navigation screens based on user permissions
 */
export const getAuthorizedScreens = (user: User | null): NavScreen[] => {
  return navigationScreens.filter(screen => screen.permission(user));
};

/**
 * Check if user can access a specific screen
 */
export const canAccessScreen = (user: User | null, screenName: string): boolean => {
  const screen = navigationScreens.find(nav => nav.name === screenName);
  if (!screen) return true; // Allow access to undefined screens
  return screen.permission(user);
};

/**
 * Get user's default landing screen based on role
 */
export const getDefaultScreen = (user: User | null): string => {
  if (!user) return 'Login';
  
  if (isSuperuser(user)) return 'Properties';
  if (isAdminOrManager(user)) return 'Dashboard';
  if (isFrontDeskOrAbove(user)) return 'Reservations';
  if (isHousekeepingStaff(user)) return 'Housekeeping';
  if (isMaintenanceStaff(user)) return 'Maintenance';
  if (isPOSStaff(user)) return 'POS';
  
  return 'Rooms'; // Default for guests or undefined roles
};

File: backend/api/permissions.py
"""
Custom permission classes for role-based access control.
"""
from rest_framework.permissions import BasePermission


class IsSuperuser(BasePermission):
    """
    Permission class that allows access only to superusers.
    Used for: Properties, Users, System Settings
    """
    def has_permission(self, request, view):
        return request.user and request.user.is_superuser


class IsAdminOrManager(BasePermission):
    """
    Permission class that allows access to Admin and Manager roles.
    Used for: Property management, Staff management
    """
    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False
        return request.user.is_superuser or request.user.role in ['ADMIN', 'MANAGER']


class IsFrontDeskOrAbove(BasePermission):
    """
    Permission class for Front Desk staff and above.
    Used for: Reservations, Guest management, Check-in/out
    """
    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False
        allowed_roles = ['ADMIN', 'MANAGER', 'FRONT_DESK']
        return request.user.is_superuser or request.user.role in allowed_roles


class IsHousekeepingStaff(BasePermission):
    """
    Permission class for Housekeeping staff.
    Used for: Housekeeping tasks, Room status updates
    """
    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False
        allowed_roles = ['ADMIN', 'MANAGER', 'HOUSEKEEPING']
        return request.user.is_superuser or request.user.role in allowed_roles


class IsMaintenanceStaff(BasePermission):
    """
    Permission class for Maintenance staff.
    Used for: Maintenance requests, Work orders
    """
    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False
        allowed_roles = ['ADMIN', 'MANAGER', 'MAINTENANCE']
        return request.user.is_superuser or request.user.role in allowed_roles


class IsAccountantOrAbove(BasePermission):
    """
    Permission class for Accountants and above.
    Used for: Financial reports, Billing, Revenue data
    """
    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False
        allowed_roles = ['ADMIN', 'MANAGER', 'ACCOUNTANT']
        return request.user.is_superuser or request.user.role in allowed_roles


class IsPOSStaff(BasePermission):
    """
    Permission class for POS staff.
    Used for: POS operations, Menu management, Order processing
    """
    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False
        allowed_roles = ['ADMIN', 'MANAGER', 'POS_STAFF']
        return request.user.is_superuser or request.user.role in allowed_roles


class IsGuest(BasePermission):
    """
    Permission class for Guest users.
    Used for: Guest portal, View own reservations, Own bills
    """
    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False
        return request.user.role == 'GUEST'


class CanViewInvoices(BasePermission):
    """
    Permission class for viewing invoices.
    Front Desk can view invoices for check-out, but not full billing access.
    """
    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False
        allowed_roles = ['ADMIN', 'MANAGER', 'ACCOUNTANT', 'FRONT_DESK']
        # Front desk can only read invoices, not create/modify
        if request.user.role == 'FRONT_DESK':
            return request.method in ['GET', 'HEAD', 'OPTIONS']
        return request.user.is_superuser or request.user.role in allowed_roles
        if not request.user or not request.user.is_authenticated:
            return False
        allowed_roles = ['ADMIN', 'MANAGER', 'POS_STAFF']
        return request.user.is_superuser or request.user.role in allowed_roles


class IsReadOnly(BasePermission):
    """
    Permission class that allows read-only access.
    Allows: GET, HEAD, OPTIONS
    Denies: POST, PUT, PATCH, DELETE
    """
    def has_permission(self, request, view):
        return request.method in ['GET', 'HEAD', 'OPTIONS']


class CanManageUsers(BasePermission):
    """
    Permission class for user management.
    Only superusers can manage users.
    """
    def has_permission(self, request, view):
        return request.user and request.user.is_superuser


class CanManageProperties(BasePermission):
    """
    Permission class for property management.
    Only superusers can create/delete properties.
    Managers can view and update their own property.
    """
    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False
        
        # Superusers can do everything
        if request.user.is_superuser:
            return True
        
        # Managers can only view (GET requests)
        if request.user.role == 'MANAGER' and request.method in ['GET', 'HEAD', 'OPTIONS']:
            return True
        
        return False
